compiler/rustc_borrowck/src/invalidation.rs

```diff
@@ -109,22 +112,36 @@ impl<'cx, 'tcx> Visitor<'tcx> for InvalidationGenerator<'cx, 'tcx> {
             TerminatorKind::SwitchInt { ref discr, switch_ty: _, targets: _ } => {
                 self.consume_operand(location, discr);
             }
-            TerminatorKind::Drop { place: drop_place, target: _, unwind: _ } => {
-                self.access_place(
-                    location,
-                    *drop_place,
-                    (AccessDepth::Drop, Write(WriteKind::StorageDeadOrDrop)),
-                    LocalMutationIsAllowed::Yes,
-                );
-            }
-            TerminatorKind::DropAndReplace {
-                place: drop_place,
-                value: ref new_value,
-                target: _,
-                unwind: _,
-            } => {
-                self.mutate_place(location, *drop_place, Deep);
-                self.consume_operand(location, new_value);
+            TerminatorKind::Drop { place: drop_place, target, unwind, is_replace } => {
+                let next_statement = if *is_replace {
+                    self.body
+                        .basic_blocks
+                        .get(*target)
+                        .expect("MIR should be complete at this point")
+                        .statements
+                        .first()
+                } else {
+                    None
+                };
+
+                match next_statement {
+                    Some(Statement { kind: StatementKind::Assign(_), source_info: _ }) => {
```

Comment:
I don't think this check is sufficient -- you are just checking that the DROP is followed by an ASSIGNMENT, but actually don't you want to check that the place being dropped or assigned is the same?

[This code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b98f6d962782c5c8a5b0389a342ebb2c) for example yields MIR like

```
bb1: {
        drop(_1) -> bb2;                 // scope 0 at [src/main.rs:4:5: 4:6](https://play.rust-lang.org/#)
    }

    bb2: {
        _2 = String::new() -> bb3;       // scope 0 at [src/main.rs:5:13: 5:26](https://play.rust-lang.org/#)
                                         // mir::Constant
                                         // + span: [src/main.rs:5:13: 5:24](https://play.rust-lang.org/#)
                                         // + literal: Const { ty: fn() -> String {String::new}, val: Value(<ZST>) }
    }
```

which I think would trigger here, but shouldn't...?

Oh, I guess it's gated by the "is-replace" flag on `Drop`, so that probably works ok. Hmm. It still seems a bit sketchy to me.

In general, I am wondering, why are the diagnostics better if we do a `mutate_place` call here? I would rather modify the way that `access_place` *reports* the error (but not the check itself). 

Also, if `is_replace` is true, shouldn't the next statement *always* be an assignment? i.e., we could replace this with an assertion.

Reply:
> Oh, I guess it's gated by the "is-replace" flag on Drop, so that probably works ok. Hmm. It still seems a bit sketchy to me.

Yep

> In general, I am wondering, why are the diagnostics better if we do a mutate_place call here? I would rather modify the way that access_place reports the error (but not the check itself).

I would be very happy to do that. I do not remember now exactly what were the differences, but it was also to try push this to be as close to what was done before as possible. I'll try to see exactly what the error messages would look like and if we can improve the `access_place` reporting.

> Also, if is_replace is true, shouldn't the next statement always be an assignment? i.e., we could replace this with an assertion.

Yes, I was not sure at first due to possible optimizations

Reply:
>In general, I am wondering, why are the diagnostics better if we do a mutate_place call here? I would rather modify the way that access_place reports the error (but not the check itself).

For the user I think it's easier to report a problem in the assignment (i.e. a mutate) rather than the implicit drop in the case of a replace. We could also report both but I think that will clutter the report unnecessarily.

Btw, is there a logical difference between `mutate_place` and `access_place` with write access? Looks like `mutate_place` is mostly a shorthand for `access_place` with a few implicit arguments. Changing the write kind from `StorageDeadOrDrop` to `Mutate` will make the error reporting in `access_place`  more user friendly for a replace and shouldn't change any logic if I understand the comment on `WriteKind` correctly.

Reply:
In the end, I've tried another option which is simply to skip the drop access check in case of a replace and rely on the assignment to report possible problems. One thing to handle separately is the check access depth, which in case of an assignment following a skipped drop access should be `AccessDepth::Drop`.